### PR TITLE
swap and rename also the link_condition check will be fixed here

### PR DESCRIPTION
**Pull request by yfzhou42**

### CODE REVIEW

File: app/EdgeOperations2d/QEC.cpp [hunk 1]
```diff
@@ -50,16 +50,19 @@ std::array<double, 10> Edge2d ::EdgeCollapse::compute_Q_v(wmtk::TriMesh::Tuple&
...
+    // Eigen::Vector3d v = m_vertex_positions[v_tuple.vid()];
+    // std::array<double, 10> Q = compute_Q_v(v_tuple);
 
-    double cost;
-    return cost;
+    // double cost;
+    // return cost;
+    assert(false);
+    return 0;
 }
```
Comment (jiangzhongshi) [inline app/EdgeOperations2d/QEC.cpp:63]: what's this function for

File: app/EdgeOperations2d/ShortestEdgeCollapse.cpp [hunk 2]
```diff
@@ -2,48 +2,38 @@
...
+        collapse_cache.v2p = m_vertex_positions[t.switch_vertex(*this).vid()];
         return true;
     }
     return false;
 }
 
-bool Edge2d::EdgeCollapse::collapse_after(const Tuple& t)
+bool Edge2d::EdgeOperations2d::collapse_after(const Tuple& t)
 {
     if (check_mesh_connectivity_validity() && t.is_valid(*this)) {
```
Comment (jiangzhongshi) [inline app/EdgeOperations2d/ShortestEdgeCollapse.cpp:21]: should not be `if`

File: app/tests/test_2d_operations.cpp [hunk 3]
```diff
@@ -173,10 +215,22 @@ TEST_CASE("shortest_edge_collapse_circle", "[test_2d_operations]")
...
+        if (cnt > 200) break;
+        if (!edge.is_valid(m)) continue;
+        if (!(edge.switch_face(m)).has_value()) {
+            REQUIRE_FALSE(m.swap_edge(edge, new_e));
+            continue;
+        }
+        REQUIRE(m.swap_edge(edge, new_e));
+        cnt++;
+    }
+    m.write_triangle_mesh("sawped.obj");
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:235]: no need to write, check its info instead

File: app/tests/test_operations.cpp [hunk 4]
```diff
@@ -49,21 +49,38 @@ TEST_CASE("edge_collapsing", "[test_operation]")
...
-    vertices[0].m_posf = Vector3f(0, 0, 0);
-    vertices[1].m_posf = Vector3f(1, 0, 0);
-    vertices[2].m_posf = Vector3f(0, 1, 0);
-    vertices[3].m_posf = Vector3f(0, 0, 1);
-    std::vector<std::array<size_t, 4>> tets = {{{0, 1, 3, 2}}};
-    std::vector<TetAttributes> tet_attrs(1);
-
-    tetwild.init(vertices.size(), tets);
-    tetwild.create_mesh_attributes(vertices, tet_attrs);
+    {
```
Comment (jiangzhongshi) [inline app/tests/test_operations.cpp:52]: Strange, these are from other pr.

File: src/wmtk/TetMesh.h [hunk 5]
```diff
@@ -366,6 +358,10 @@ class TetMesh
     virtual bool tetrahedron_invariant(const Tuple& t) { return true; }
 
     virtual void resize_attributes(size_t v, size_t e, size_t f, size_t t) {}
+    virtual void move_face_attribute(size_t from, size_t to) {}
```
Comment (jiangzhongshi) [inline src/wmtk/TetMesh.h:361]: These are from the compact PR.

File: app/tests/test_operations.cpp (hunk 4 above)
Comment (jiangzhongshi) [reply app/tests/test_operations.cpp:52]: Do not copy file from other branches. Merge instead: it creates a lot of conflicts this way.

File: app/EdgeOperations2d/QEC.cpp (hunk 1 above)
Comment (yfzhou42) [reply app/EdgeOperations2d/QEC.cpp:63]: That was for the implementation of qec

Comment (yfzhou42) [reply app/EdgeOperations2d/QEC.cpp:63]: updating the cost of each edge

File: app/tests/test_operations.cpp (hunk 4 above)
Comment (yfzhou42) [reply app/tests/test_operations.cpp:52]: I did not. I merged from some branches Teseo or Daniele worked on previous

File: app/EdgeOperations2d/ShortestEdgeCollapse.cpp [hunk 6]
```diff
@@ -2,48 +2,35 @@
...
-    if (new_vid < m_vertex_positions.size())
-        m_vertex_positions[new_vid] = new_position;
-    else
-        m_vertex_positions.emplace_back(new_position);
-}
 
-bool Edge2d::EdgeCollapse::collapse_before(const Tuple& t)
+bool Edge2d::EdgeOperations2d::collapse_before(const Tuple& t)
 {
     if (check_link_condition(t) && check_manifold(t)) {
```
Comment (teseoch) [inline app/EdgeOperations2d/ShortestEdgeCollapse.cpp:11]: this should go in the super class

File: src/wmtk/TriMesh.cpp [hunk 7]
```diff
@@ -612,15 +646,69 @@ size_t TriMesh::get_next_empty_slot_v()
 
 bool TriMesh::check_manifold(const Tuple& t) const
 {
-    auto v1_conn_tris = m_vertex_connectivity[t.vid()].m_conn_tris;
-    auto v2_conn_tris = m_vertex_connectivity[t.switch_vertex(*this).vid()].m_conn_tris;
+    // check collapse to one edge
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:649]: Here should be checked by link condition?

File: app/tests/test_2d_operations.cpp [hunk 8]
```diff
@@ -79,42 +76,83 @@ TEST_CASE("shortest_edge_collapse_boundary_edge", "[test_2d_operations]")
...
+            for (int j = 0; j < 3; j++) tri[i][j] = (size_t)F(i, j);
+        }
+        EdgeOperations2d m(v);
+        m.create_mesh(V.rows(), tri);
+        REQUIRE(m.check_mesh_connectivity_validity());
+        REQUIRE(m.collapse_shortest(100));
+        REQUIRE(m.get_vertices().size() == 4);
+
+        REQUIRE(m.get_faces().size() == 4);
+        m.write_triangle_mesh("collapsed.obj");
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:140]:
```suggestion
```

File: app/tests/test_2d_operations.cpp [hunk 9]
```diff
@@ -123,19 +161,18 @@ TEST_CASE("shortest_edge_collapse_on_mesh", "[test_2d_operations]")
     for (int i = 0; i < F.rows(); i++) {
         for (int j = 0; j < 3; j++) tri[i][j] = (size_t)F(i, j);
     }
-    EdgeCollapse m(v);
+    EdgeOperations2d m(v);
     m.create_mesh(V.rows(), tri);
     REQUIRE(m.check_mesh_connectivity_validity());
     std::cout << " is it mesh passed " << std ::endl;
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:167]:
```suggestion
```

File: app/tests/test_2d_operations.cpp [hunk 10]
```diff
@@ -123,19 +161,18 @@ TEST_CASE("shortest_edge_collapse_on_mesh", "[test_2d_operations]")
...
-    REQUIRE(m.collapse_shortest(5));
+    REQUIRE(m.collapse_shortest(50));
     m.write_triangle_mesh("collapsed.obj");
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:169]:
```suggestion
```

File: app/tests/test_2d_operations.cpp [hunk 11]
```diff
@@ -151,15 +188,15 @@ TEST_CASE("shortest_edge_collapse_octocat", "[test_2d_operations]")
     for (int i = 0; i < F.rows(); i++) {
         for (int j = 0; j < 3; j++) tri[i][j] = (size_t)F(i, j);
     }
-    EdgeCollapse m(v);
+    EdgeOperations2d m(v);
     m.create_mesh(V.rows(), tri);
     REQUIRE(m.check_mesh_connectivity_validity());
     std::cout << " is it mesh passed " << std ::endl;
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:194]:
```suggestion
```

File: app/tests/test_2d_operations.cpp [hunk 12]
```diff
@@ -178,10 +215,22 @@ TEST_CASE("shortest_edge_collapse_circle", "[test_2d_operations]")
...
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:221]:
```suggestion
```

File: app/tests/test_2d_operations.cpp [hunk 13]
```diff
@@ -151,15 +188,15 @@ TEST_CASE("shortest_edge_collapse_octocat", "[test_2d_operations]")
...
     REQUIRE(m.collapse_shortest(100));
     m.write_triangle_mesh("collapsed.obj");
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:196]:
```suggestion
```

File: src/wmtk/TriMesh.cpp [hunk 14]
```diff
@@ -440,7 +465,16 @@ void TriMesh::consolidate_mesh()
 
     resize_attributes(v_cnt, 3 * t_cnt, t_cnt);
 
-    check_mesh_connectivity_validity();
+    // Resize user class attributes
+    resize_attributes(
+        m_vertex_connectivity.size(),
+        m_tri_connectivity.size() * 3,
+        m_tri_connectivity.size());
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:472]: Duplicates?

File: src/wmtk/TriMesh.cpp [hunk 15]
```diff
@@ -612,15 +646,69 @@ size_t TriMesh::get_next_empty_slot_v()
...
+// link check, prerequisite for edge collapse
+bool wmtk::TriMesh::check_link_condition(const Tuple& edge) const
+{
+    assert(edge.is_valid(*this));
+    size_t vid1 = edge.vid();
+    size_t vid2 = switch_vertex(edge).vid();
+    auto vid1_ring = get_one_ring_edges_for_vertex(edge);
+    auto vid2_ring = get_one_ring_edges_for_vertex(switch_vertex(edge));
+
+    size_t dummy = std::numeric_limits<unsigned>::max();
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:690]:
```suggestion
    size_t dummy = std::numeric_limits<size_t>::max();
```

File: tests/2d_tests.cpp [hunk 16]
```diff
@@ -180,13 +181,13 @@ TriMesh::Tuple double_switch_face(TriMesh::Tuple t, TriMesh& m)
     return t_after;
 }
 
-bool tuple_equal(TriMesh::Tuple t1, TriMesh::Tuple t2)
+bool tuple_equal(const TriMesh& m, TriMesh::Tuple t1, TriMesh::Tuple t2)
 {
-    if (t1.vid() != t2.vid()) std::cout << "vids : " << t1.vid() << " " << t2.vid() << std::endl;
-    if (t1.eid() != t2.eid()) std::cout << "eids : " << t1.eid() << " " << t2.eid() << std::endl;
-    if (t1.fid() != t2.fid()) std::cout << "fids : " << t1.fid() << " " << t2.fid() << std::endl;
+    if (t1.vid() != t2.vid()) throw "vids of two tuples not equal";
```
Comment (jiangzhongshi) [inline tests/2d_tests.cpp:186]: return false?

File: tests/2d_tests.cpp [hunk 17]
```diff
@@ -297,54 +298,157 @@ TEST_CASE("vertex_edge switches equals indentity", "[test_operation]")
...
+        std::vector<size_t> m_indices;
+        for (auto edge : m.get_one_ring_edges_for_vertex(new_e)) {
+            m_indices.push_back(edge.vid());
+        }
+        vector_unique(m_indices);
+        vector_print(m_indices);
+        std::vector<size_t> tru_indices = {1, 3, 4, 5};
+        REQUIRE(std::equal(m_indices.begin(), m_indices.end(), tru_indices.begin()));
+    }
+    SECTION("sawp_on_boundary_edge")
```
Comment (jiangzhongshi) [inline tests/2d_tests.cpp:447]:
```suggestion
    SECTION("swap_on_boundary_edge")
```

File: tests/2d_tests.cpp [hunk 18]
```diff
@@ -297,54 +298,157 @@ TEST_CASE("vertex_edge switches equals indentity", "[test_operation]")
...
+        REQUIRE(new_e.is_valid(m));
+        REQUIRE(m.vert_capacity() == 8);
+        REQUIRE(m.tri_capacity() == 8);
+
+        std::vector<size_t> m_indices;
+        for (auto edge : m.get_one_ring_edges_for_vertex(new_e)) {
+            m_indices.push_back(edge.vid());
+        }
+        vector_unique(m_indices);
+        vector_print(m_indices);
```
Comment (jiangzhongshi) [inline tests/2d_tests.cpp:443]:
```suggestion
```

File: tests/2d_tests.cpp [hunk 19]
```diff
@@ -180,13 +181,13 @@ TriMesh::Tuple double_switch_face(TriMesh::Tuple t, TriMesh& m)
...
 
-bool tuple_equal(TriMesh::Tuple t1, TriMesh::Tuple t2)
+bool tuple_equal(const TriMesh& m, TriMesh::Tuple t1, TriMesh::Tuple t2)
 {
-    if (t1.vid() != t2.vid()) std::cout << "vids : " << t1.vid() << " " << t2.vid() << std::endl;
-    if (t1.eid() != t2.eid()) std::cout << "eids : " << t1.eid() << " " << t2.eid() << std::endl;
-    if (t1.fid() != t2.fid()) std::cout << "fids : " << t1.fid() << " " << t2.fid() << std::endl;
+    if (t1.vid() != t2.vid()) throw "vids of two tuples not equal";
+    if (t1.eid(m) != t2.eid(m)) throw "eid of two tuples not equal";
+    if (t1.fid() != t2.fid()) throw "fid of two tuples not equal";
```
Comment (jiangzhongshi) [inline tests/2d_tests.cpp:188]:
```suggestion
```

File: src/wmtk/TriMesh.cpp (hunk 14 above)
Comment (jiangzhongshi) [reply src/wmtk/TriMesh.cpp:472]: Done.

File: tests/2d_tests.cpp [hunk 20]
```diff
@@ -297,54 +295,156 @@ TEST_CASE("vertex_edge switches equals indentity", "[test_operation]")
...
+    {
+        std::vector<std::array<size_t, 3>> tris = {{{0, 1, 2}}};
+        m.create_mesh(3, tris);
 
-// these roll back tests must have more than 2 triangles, since by design when there are only two
-// incident triangles the edge cannot be collapsed
-// this check should be doen in pre collapse check
+        TriMesh::Tuple edge(0, 2, 0, m);
+        assert(edge.is_valid(m));
+        REQUIRE(m.check_link_condition(edge));
```
Comment (jiangzhongshi) [inline tests/2d_tests.cpp:321]: The edge should violate link condition.

**Comment by jiangzhongshi** [comment]
I put in some additional fixes. Apart from the link condition, which can be added in a future PR, this is good to go.

**Review by jiangzhongshi — APPROVED** [review]

**Pull request merged**